### PR TITLE
Add attributes for required members

### LINKED ISSUE #64248: [API Proposal]: RequiredMemberAttribute and NoRequiredMembersAttribute
### Background and motivation

As part of supporting C# 11's [Required Members](https://github.com/dotnet/csharplang/blob/main/proposals/required-members.md) feature, we need two new attributes: `RequiredMemberAttribute` and `NoRequiredMembersAttribute`. They work as follows:

* `RequiredMemberAttribute` will be applied to every type (class or struct) that defines required members, and to all the members (fields or properties) that are required by that type.
* `NoRequiredMembersAttribute` will be applied to a constructor of a type that has required members, and then users of that constructor will not be required to set any members of the type when calling it. Think copy constructors for records: this is how they will communicate that calling the constructor does not require setting any members, because all members will have been set by the copy constructor itself.

### API Proposal

```C#
namespace System.Runtime.CompilerServices
{
    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Struct | AttributeTargets.Field | AttributeTargets.Property)]
    public sealed class RequiredMemberAttribute : Attribute { }
}
namespace System.Diagnostics.CodeAnalysis
{
    [AttributeUsage(AttributeTargets.Constructor)]
    public sealed class NoRequiredMembersAttribute  : Attribute { }
}
```

### API Usage

`RequiredMemberAttribute` will not be manually usable. It will be an error to apply it to C# code by hand.

A copy constructor for a type that has required members might look like this:

```cs
public class C
{
    public required int Property { get; init; }
    public C() {}

    [NoRequiredMembers]
    public C(C c) => this.Property = c.Property;
}


// Usage
var c1 = new C { Property = 1 };
var c2 = new C(c1); // No error, constructor waives requirements
```

### Alternative Designs

I am open to alternative naming schemes. We had also considered a version of `RequiredMemberAttribute` where all the members were listed by name as a string argument to the constructor, but we redesigned after some community feedback and consultation with Wrighton to make sure that more attributes wouldn't break perf.

### Risks

_No response_

### PR DESCRIPTION
**Pull request by 333fred**
Closes https://github.com/dotnet/runtime/issues/64248.

Relates to test plan https://github.com/dotnet/roslyn/issues/57046

### CODE REVIEW

**Comment by stephentoub** [comment]
This also needs:
- To add these files to the csproj: https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems
- XML comments on the APIs to seed the documentation
- To add these to the refs: https://github.com/dotnet/runtime/blob/main/src/libraries/System.Runtime/ref/System.Runtime.cs
- Minimal tests, ala https://github.com/dotnet/runtime/blob/main/src/libraries/System.Runtime/tests/System/Diagnostics/CodeAnalysis/ConstantExpectedAttributeTests.cs.  Given there are no members, the test will just validate the instance can be constructed.

Also the ifdef for corelib is only necessary if we expect these files to be included into other assemblies beyond corelib.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/SetsRequiredMembersAttribute.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#nullable enable
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/SetsRequiredMembersAttribute.cs:4]: These #nullable enables aren't necessary as we enable it at the project level (and there's nothing in the APIs that's impacted).

Comment (333fred) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/SetsRequiredMembersAttribute.cs:4]: Ah. I was just copying from other attributes in the namespace.

**Comment by 333fred** [comment]
> Also the ifdef for corelib is only necessary if we expect these files to be included into other assemblies beyond corelib.

I somewhat expect that this attribute will be copied by customers, so I'd prefer to keep them.

File: src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems [hunk 2]
```diff
@@ -745,6 +746,7 @@
     <Compile Include="$(MSBuildThisFileDirectory)System\Runtime\CompilerServices\PoolingAsyncValueTaskMethodBuilder.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)System\Runtime\CompilerServices\PoolingAsyncValueTaskMethodBuilderT.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)System\Runtime\CompilerServices\PreserveBaseOverridesAttribute.cs" />
+    <Compile Include="$(MSBuildThisFileDirectory)System\Runtime\CompilerServices\RequiredMembersAttribute.cs" />
```
Comment (333fred) [inline src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems:749]:
```suggestion
    <Compile Include="$(MSBuildThisFileDirectory)System\Runtime\CompilerServices\RequiredMemberAttribute.cs" />
```

File: src/libraries/System.Runtime/ref/System.Runtime.cs [hunk 3]
```diff
@@ -9180,6 +9180,11 @@ public sealed partial class RequiresUnreferencedCodeAttribute : System.Attribute
         public string Message { get { throw null; } }
         public string? Url { get { throw null; } set { } }
     }
+    [AttributeUsage(AttributeTargets.Constructor, AllowMultiple = false, Inherited = false)]
+    public sealed class SetsRequiredMembersAttribute : Attribute
```
Comment (stephentoub) [inline src/libraries/System.Runtime/ref/System.Runtime.cs:9184]:
```suggestion
    public sealed class SetsRequiredMembersAttribute : System.Attribute
```

File: src/libraries/System.Runtime/ref/System.Runtime.cs [hunk 4]
```diff
@@ -9180,6 +9180,11 @@ public sealed partial class RequiresUnreferencedCodeAttribute : System.Attribute
         public string Message { get { throw null; } }
         public string? Url { get { throw null; } set { } }
     }
+    [AttributeUsage(AttributeTargets.Constructor, AllowMultiple = false, Inherited = false)]
```
Comment (stephentoub) [inline src/libraries/System.Runtime/ref/System.Runtime.cs:9183]:
```suggestion
    [System.AttributeUsage(System.AttributeTargets.Constructor, AllowMultiple = false, Inherited = false)]
```

File: src/libraries/System.Runtime/ref/System.Runtime.cs [hunk 5]
```diff
@@ -13194,6 +13199,11 @@ public sealed partial class PreserveBaseOverridesAttribute : System.Attribute
     {
         public PreserveBaseOverridesAttribute() { }
     }
+    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Struct | AttributeTargets.Field | AttributeTargets.Property, AllowMultiple = false, Inherited = false)]
```
Comment (stephentoub) [inline src/libraries/System.Runtime/ref/System.Runtime.cs:13202]:
```suggestion
    [System.AttributeUsage(System.AttributeTargets.Class | System.AttributeTargets.Struct | System.AttributeTargets.Field | System.AttributeTargets.Property, AllowMultiple = false, Inherited = false)]
```

File: src/libraries/System.Runtime/ref/System.Runtime.cs [hunk 6]
```diff
@@ -13194,6 +13199,11 @@ public sealed partial class PreserveBaseOverridesAttribute : System.Attribute
...
+    public sealed class RequiredMemberAttribute : Attribute
```
Comment (stephentoub) [inline src/libraries/System.Runtime/ref/System.Runtime.cs:13203]:
```suggestion
    public sealed class RequiredMemberAttribute : System.Attribute
```

File: src/libraries/System.Runtime/tests/System/Runtime/CompilerServices/AttributesTests.cs [hunk 7]
```diff
@@ -320,5 +320,11 @@ public static void InterpolatedStringHandlerArgumentAttributeTests(string firstP
             Assert.Equal(firstParameterName, attr2.Arguments[0]);
             Assert.Equal("param2", attr2.Arguments[1]);
         }
+
+        [Fact]
+        public static void RequiredMemberAttributeTests()
+        {
+            new RequiredMemberAttribute();
+        }
```
Comment (stephentoub) [inline src/libraries/System.Runtime/tests/System/Runtime/CompilerServices/AttributesTests.cs:328]: The two test files will need to be added to https://github.com/dotnet/runtime/blob/main/src/libraries/System.Runtime/tests/System.Runtime.Tests.csproj as well.

**Review by stephentoub — APPROVED** [review]
Thanks.

File: src/libraries/System.Runtime/tests/System/Runtime/CompilerServices/AttributesTests.cs (hunk 7 above)
Comment (333fred) [reply src/libraries/System.Runtime/tests/System/Runtime/CompilerServices/AttributesTests.cs:328]: This one already is, but I'll add the other.

**Review by stephentoub — APPROVED** [review]
LGTM. Thanks.

**Pull request merged**